fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/ComparatorCursor.java

```diff
@@ -0,0 +1,142 @@
+
+/**
+ * A cursor that implements a comparison of matching elements from a set of cursors all of whom are ordered compatibly.
+ * For each record returned by the sub-cursors, this cursor will ensure that they are the same (by using the comparison
+ * function). The cursor will log and continue if the records are not the same or if one of the sub-cursors ends
+ * prematurely.
+ * The cursor defines a "reference cursor" whose results are assumed to be correct. Values are compared to this
+ * cursor's results, and the reference cursor's results are the ones that are actually returned.
+ *
+ * @param <T> the type of elements returned by the cursor
+ */
+@API(API.Status.EXPERIMENTAL)
+public class ComparatorCursor<T> extends ComparatorCursorBase<T, T> {
+
+    private ComparatorCursor(@Nonnull List<KeyedMergeCursorState<T>> cursorStates,
+                             @Nullable FDBStoreTimer timer,
+                             final int referencePlanIndex,
+                             @Nonnull final Supplier<String> planStringSupplier,
+                             @Nonnull final Supplier<Integer> planHashSupplier) {
+        super(cursorStates, timer, referencePlanIndex, planStringSupplier, planHashSupplier);
+    }
+
+    @Override
+    @Nonnull
+    protected T getNextResult(@Nonnull List<KeyedMergeCursorState<T>> cursorStates) {
+        return getReferenceState(cursorStates).getResult().get();
+    }
+
+    /**
+     * Create a comparator cursor from several compatibly-ordered cursors.
+     * As its comparison key function, it will evaluate the provided comparison key
+     * on each record from each cursor. This otherwise behaves exactly the same way
+     * as the overload of this function that takes a function to extract a comparison
+     * key.
+     *
+     * @param <M> the type of the Protobuf record elements of the cursor
+     * @param <S> the type of record wrapping a record of type <code>M</code>
+     * @param store record store from which records will be fetched
+     * @param comparisonKey the key expression used to compare records from different cursors
+     * @param cursorFunctions a list of functions to produce {@link RecordCursor}s from a continuation
+     * @param continuation any continuation from a previous scan
+     * @param referencePlanIndex the index of the reference cursor in the given list of sub-cursors
+     * @param planStringSupplier function to return the cursor plan's hash (for logging purposes)
+     * @param planHashSupplier function to return the cursor plan's string (for logging purposes)
+     *
+     * @return a cursor containing same records in all child cursors
+     *
+     * @see #create(Function, List, byte[], FDBStoreTimer, int, Supplier, Supplier)
+     */
+    @Nonnull
+    public static <M extends Message, S extends FDBRecord<M>> ComparatorCursor<S> create(
+            @Nonnull FDBRecordStoreBase<M> store,
+            @Nonnull KeyExpression comparisonKey,
```

Comment:
Can you `Verify` that we are not handing comparison keys that use nested repeateds -- look for `createdDuplicates()` on `KeyExpression`

Reply:
Sure. Just to make sure I understand: Is the problem here that the key itself can evaluate to repeated field?
You also mentioned that in cases sub-plans have predicates that operate on repeated fields then we can have duplicate results (and they can differ in their distinct-ness), and that in this case comparison may fail, but this is a different issue, right?

Reply:
If `createsDuplicates()` returns `true` a particular index may actually return multiple key, value pairs for a document (which may flow out of order -- i.e. not next to each other). I just want us to not extend this logic to that sort of complication just yet. The compared correctness would depend on the index chosen which seems to be wrong. In the future we may relax this to an execution-time check, if we really that there is only ever one duplicate. That is the case for maps.

Reply:
Agree. My question is what do you propose we do to avoid that. The situation currently is that the query execution will fail since there is a validation inside the `MergeCursor` to check that the number of results of applying the `comparisonKey` has exactly one result.
What I recommend we do is, in a future PR, change the `MergeCursor` to allow multiple records per key and be able to return them correctly. For now, we can either:
1. Fail the creation of the plan
2. Allow the plan to be created but do not create a comparison plan
I don't like option 2 since it will hide the problem and at runtime it will look like the comparison ran without failure.
If you agree, I can implement 1 from above for now.